### PR TITLE
Deprecation below 7.2.5

### PR DESCRIPTION
**Pull request by SamRemis**
This pull request is a work in progress and will be added to over time.  This will be merged before the release on Tuesday, August 15th, 2023 and will be a minor version bump that ends support for php versions below 7.2.5.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ jobs:
     strategy:
       #for each of the following versions of PHP, with and without --prefer-lowest
       matrix:
-        php-versions: ['5.5', '5.6', '7.0', '7.1', '7.2', '7.3', '7.4', '8.0', '8.1', '8.2']
+        php-versions: ['7.2.5', '7.3', '7.4', '8.0', '8.1', '8.2']
```
Comment (GrahamCampbell) [inline .github/workflows/tests.yml:19]: One should test on 7.2.5 and also 7.2 (7.2 is not 7.2.0, but the latest 7.2.x version)

File: src/HashingStream.php [hunk 2]
```diff
@@ -50,14 +50,13 @@ public function read($length)
...
     {
         if ($offset === 0) {
             $this->hash->reset();
-            return $this->stream->seek($offset);
+            $this->stream->seek($offset);
         }
-
         // Seeking arbitrarily is not supported.
-        return false;
+        return;
```
Comment (GrahamCampbell) [inline src/HashingStream.php:60]: This function can be simplified to:

```php
public function seek($offset, $whence = SEEK_SET): void
{
    // Seeking arbitrarily is not supported.
    if ($offset !== 0) {
        return;
    }

    $this->hash->reset();
    $this->stream->seek($offset);
}
```

File: src/Multipart/AbstractUploadManager.php [hunk 3]
```diff
@@ -219,9 +219,8 @@ abstract protected function getCompleteParams();
      * Based on the config and service-specific workflow info, creates a
      * `Promise` for an `UploadState` object.
      *
```
Comment (GrahamCampbell) [inline src/Multipart/AbstractUploadManager.php:221]: This blank line should be removed.

File: src/Rds/AuthTokenGenerator.php [hunk 4]
```diff
@@ -47,8 +47,7 @@ public function createToken($endpoint, $region, $username, $lifetime = 15)
     {
         if (!is_numeric($lifetime) || $lifetime > 15 || $lifetime <= 0) {
             throw new \InvalidArgumentException(
-                "Lifetime must be a positive number less than or equal to 15, was {$lifetime}",
-                null
```
Comment (GrahamCampbell) [inline src/Rds/AuthTokenGenerator.php:51]: this fix doesn't seem related to the PHP upgrade

File: src/S3/StreamWrapper.php [hunk 5]
```diff
@@ -109,9 +109,9 @@ class StreamWrapper
      */
     public static function register(
         S3ClientInterface $client,
-        $protocol = 's3',
+                          $protocol = 's3',
```
Comment (GrahamCampbell) [inline src/S3/StreamWrapper.php:112]: this change is incorrect

File: src/S3/StreamWrapper.php [hunk 6]
```diff
@@ -109,9 +109,9 @@ class StreamWrapper
...
         CacheInterface $cache = null,
-        $v2Existence = false
+                          $v2Existence = false
```
Comment (GrahamCampbell) [inline src/S3/StreamWrapper.php:114]: this change is incorrect

File: src/S3/StreamWrapper.php [hunk 7]
```diff
@@ -569,9 +569,9 @@ public function rename($path_from, $path_to)
             );
             // Delete the original object
             $this->getClient()->deleteObject([
-                'Bucket' => $partsFrom['Bucket'],
-                'Key'    => $partsFrom['Key']
-            ] + $options);
+                    'Bucket' => $partsFrom['Bucket'],
```
Comment (GrahamCampbell) [inline src/S3/StreamWrapper.php:572]: this change is incorrect

File: tests/MiddlewareTest.php [hunk 8]
```diff
@@ -28,7 +28,7 @@ class MiddlewareTest extends TestCase
 {
     public function set_up()
     {
-        \GuzzleHttp\Promise\queue()->run();
+        \GuzzleHttp\Promise\Utils::queue()->run();
```
Comment (GrahamCampbell) [inline tests/MiddlewareTest.php:31]: This fix should be moved into master, already, since `GuzzleHttp\Promise\queue` is deprecated on 1.x, too.

File: tests/MiddlewareTest.php [hunk 9]
```diff
@@ -40,7 +40,7 @@ public function testCanTapIntoHandlerList()
         }));
         $handler = $list->resolve();
         $handler(new Command('foo'), new Request('GET', 'http://exmaple.com'));
-        Promise\queue()->run();
+        \GuzzleHttp\Promise\Utils::queue()->run();
```
Comment (GrahamCampbell) [inline tests/MiddlewareTest.php:43]: This fix should be moved into master, already.

File: tests/MiddlewareTest.php [hunk 10]
```diff
@@ -55,7 +55,7 @@ public function testWrapsWithRetryMiddleware()
...
```
Comment (GrahamCampbell) [inline tests/MiddlewareTest.php:58]: This fix should be moved into master, already.

File: tests/MiddlewareTest.php [hunk 11]
```diff
@@ -104,7 +104,7 @@ public function testAddsSigner()
         $list->appendSign(Middleware::signer($creds, Aws\constantly($signature)));
         $handler = $list->resolve();
         $handler(new Command('foo'), new Request('GET', 'http://exmaple.com'));
-        Promise\queue()->run();
+        \GuzzleHttp\Promise\Utils::queue()->run();
```
Comment (GrahamCampbell) [inline tests/MiddlewareTest.php:107]: This fix should be moved into master, already.

File: tests/MiddlewareTest.php [hunk 12]
```diff
@@ -247,7 +247,7 @@ public function testExtractsSourceFileIntoBody()
             'Key'        => 'key',
             'SourceFile' => __FILE__
         ]), new Request('PUT', 'http://foo.com'));
-        Promise\queue()->run();
+        \GuzzleHttp\Promise\Utils::queue()->run();
```
Comment (GrahamCampbell) [inline tests/MiddlewareTest.php:250]: This fix should be moved into master, already.

File: tests/MiddlewareTest.php [hunk 13]
```diff
@@ -261,7 +261,7 @@ public function testAppliesHistory()
         $req = new Request('GET', 'http://www.foo.com');
         $cmd = new Command('foo');
         $handler($cmd, $req);
-        Promise\queue()->run();
+        \GuzzleHttp\Promise\Utils::queue()->run();
```
Comment (GrahamCampbell) [inline tests/MiddlewareTest.php:264]: This fix should be moved into master, already.

File: tests/MockHandlerTest.php [hunk 14]
```diff
@@ -91,7 +90,7 @@ function ($reason) use (&$thens) {
         $cmd = new Command('foo');
         $h($cmd, $request);
         $h($cmd, $request);
-        Promise\queue()->run();
+        \GuzzleHttp\Promise\Utils::queue()->run();
```
Comment (GrahamCampbell) [inline tests/MockHandlerTest.php:93]: This fix should be moved into master, already.

File: tests/TraceMiddlewareTest.php [hunk 15]
```diff
@@ -47,7 +47,7 @@ public function testEmitsDebugInfo()
         $command = new Command('foo', ['a' => '1', 'b' => '2']);
         $request = new Request('GET', 'http://foo.com');
         $handler($command, $request);
-        Promise\queue()->run();
+        \GuzzleHttp\Promise\Utils::queue()->run();
```
Comment (GrahamCampbell) [inline tests/TraceMiddlewareTest.php:50]: This fix should be moved into master, already.

File: tests/TraceMiddlewareTest.php [hunk 16]
```diff
@@ -83,7 +83,7 @@ public function testTracksExceptions()
         $command = new Command('foo');
         $request = new Request('GET', 'http://foo.com');
         $handler($command, $request);
-        Promise\queue()->run();
+        \GuzzleHttp\Promise\Utils::queue()->run();
```
Comment (GrahamCampbell) [inline tests/TraceMiddlewareTest.php:86]: This fix should be moved into master, already.

File: tests/TraceMiddlewareTest.php [hunk 17]
```diff
@@ -118,7 +118,7 @@ public function testTracksAwsSpecificExceptions()
...
```
Comment (GrahamCampbell) [inline tests/TraceMiddlewareTest.php:121]: This fix should be moved into master, already.

File: composer.json [hunk 18]
```diff
@@ -16,16 +16,16 @@
...
     },
     "require": {
-        "php": ">=5.5",
-        "guzzlehttp/guzzle": "^6.5.8 || ^7.4.5",
-        "guzzlehttp/psr7": "^1.9.1 || ^2.4.5",
-        "guzzlehttp/promises": "^1.4.0",
+        "php": ">=7.2.5",
+        "guzzlehttp/guzzle": "^7.4.5",
+        "guzzlehttp/psr7": "^2.4.5",
+        "guzzlehttp/promises": "^2.0",
```
Comment (GrahamCampbell) [inline composer.json:22]: we don't need to drop 1.x here. we can have `^1.5.3 || ^2.0`.

File: composer.json [hunk 19]
```diff
@@ -16,16 +16,16 @@
         "issues": "https://github.com/aws/aws-sdk-php/issues"
     },
     "require": {
-        "php": ">=5.5",
-        "guzzlehttp/guzzle": "^6.5.8 || ^7.4.5",
-        "guzzlehttp/psr7": "^1.9.1 || ^2.4.5",
-        "guzzlehttp/promises": "^1.4.0",
+        "php": ">=7.2.5",
+        "guzzlehttp/guzzle": "^7.4.5",
+        "guzzlehttp/psr7": "^2.4.5",
```
Comment (GrahamCampbell) [inline composer.json:21]: we don't need to drop 1.x here. we can have `^1.9.1 || ^2.4.5`. having extra return types in your implementations is still compatible with v1.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/S3/ObjectCopier.php [hunk 20]
```diff
@@ -79,7 +79,7 @@ public function __construct(
      *
      * @return Coroutine
      */
-    public function promise()
+    public function promise(): \GuzzleHttp\Promise\PromiseInterface
```
Comment (stobrien89) [inline src/S3/ObjectCopier.php:82]: can we add a use statement and shorten the return type?

File: tests/MiddlewareTest.php (hunk 8 above)
Comment (SamRemis) [reply tests/MiddlewareTest.php:31]: Thank you Graham, we merged this today in #2711; will address the other comments soon

**Review by stobrien89 — APPROVED** [review]

File: build/packager.php [hunk 21]
```diff
@@ -32,8 +32,8 @@
 $autoloaderContents = [
     'Aws/functions.php',
     'GuzzleHttp/functions_include.php',
-    'GuzzleHttp/Psr7/functions_include.php',
-    'GuzzleHttp/Promise/functions_include.php',
+    'GuzzleHttp/Psr7/Utils.php',
```
Comment (GrahamCampbell) [inline build/packager.php:35]: It is not necessary to include these two utils files.

Comment (stobrien89) [reply build/packager.php:35]: @GrahamCampbell,

Happy to revert this if you wanted to give the PR another once-over.

Comment (GrahamCampbell) [reply build/packager.php:35]: It's not a bug, just unnecessary to include these classes explicitly, since they already get autoloaded like the others. The original code was only necessary to include functions.

Comment (stobrien89) [reply build/packager.php:35]: We'll keep that in mind and will probably add something to our backlog for this.  If you notice anything else, definitely let us know.
